File: 109. Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        return build(head, NULL);
    }
    TreeNode* build(ListNode* left, ListNode* right)
    {
        if (left == right) return NULL;
        ListNode* mid = findmid(left, right);
        TreeNode* root = new TreeNode(mid->val);
        root->left = build(left, mid);
        root->right = build(mid->next, right);
        return root;
    }
    ListNode* findmid(ListNode* left, ListNode* right)
    {
        ListNode* p1 = left;
        ListNode* p2 = left;
        while (p2 != right)
        {
            p2 = p2->next;
            if (p2 != right)
            {
                p2 = p2->next;
                p1 = p1->next;
            }
        }
        return p1;
    }
};

File: 109. Convert Sorted List to Binary Search Tree.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        return self.build(head, None)
        
    def build(self, left, right):
        if left == right:
            return None
        mid = self.findmid(left, right)
        root = TreeNode(mid.val)
        root.left = self.build(left, mid)
        root.right = self.build(mid.next, right)
        return root
    
    def findmid(self, left, right):
        p1 = left
        p2 = left
        while p2 != right:
            p2 = p2.next
            if p2 != right:
                p2 = p2.next
                p1 = p1.next
        return p1
